File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (!preorder.length || !inorder.length) return null

    const rootVal = preorder[0]
    const root = new TreeNode(rootVal)

    const mid = inorder.indexOf(rootVal)

    const leftInorder = inorder.slice(0, mid)
    const rightInorder = inorder.slice(mid + 1)

    const leftPreorder = preorder.slice(1, mid + 1)
    const rightPreorder = preorder.slice(mid + 1)

    root.left = buildTree(leftPreorder, leftInorder)
    root.right = buildTree(rightPreorder, rightInorder)

    return root
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorder_indexes_map = {}
        for i, val in enumerate(inorder):
            inorder_indexes_map[val] = i

        preorder_index = [0]

        def build(left: int, right: int) -> Optional[TreeNode]:
            if left > right:
                return None

            val = preorder[preorder_index[0]]

            node = TreeNode(val)

            inorder_index = inorder_indexes_map[val]

            preorder_index[0] += 1

            node.left = build(left, inorder_index - 1)

            node.right = build(inorder_index + 1, right)

            return node

        return build(0, len(inorder) - 1)